binding/tfjs_backend.cc

```diff
@@ -17,14 +17,379 @@
 
 #include "tfjs_backend.h"
 
+#include "tf_auto_tensor.h"
 #include "tfe_auto_op.h"
-#include "tfe_utils.h"
 #include "utils.h"
 
 #include <memory>
+#include <set>
+#include <string>
 
 namespace tfnodejs {
 
+// Used to hold strings beyond the lifetime of a JS call.
+static std::set<std::string> ATTR_NAME_SET;
+
+TFE_TensorHandle *CreateTFE_TensorHandleFromTypedArray(
+    napi_env env, int64_t *shape, uint32_t shape_length, TF_DataType dtype,
+    napi_value typed_array_value) {
+  napi_status nstatus;
+
+  napi_typedarray_type array_type;
+  size_t array_length;
+  void *array_data;
+  nstatus =
+      napi_get_typedarray_info(env, typed_array_value, &array_type,
+                               &array_length, &array_data, nullptr, nullptr);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  // Double check the underlying TF_Tensor type matches the supplied
+  // typed-array.
+  size_t width = 0;
+  switch (array_type) {
+    case napi_float32_array:
+      if (dtype != TF_FLOAT) {
+        NAPI_THROW_ERROR(env, "Tensor type does not match Float32Array");
+        return nullptr;
+      }
+      width = sizeof(float);
+      break;
+    case napi_int32_array:
+      if (dtype != TF_INT32) {
+        NAPI_THROW_ERROR(env, "Tensor type does not match Int32Array");
+        return nullptr;
+      }
+      width = sizeof(int32_t);
+      break;
+    case napi_uint8_array:
+      if (dtype != TF_BOOL) {
+        NAPI_THROW_ERROR(env, "Tensor type does not match Uint8Array");
+        return nullptr;
+      }
+      width = sizeof(uint8_t);
+      break;
+    default:
+      REPORT_UNKNOWN_TYPED_ARRAY_TYPE(env, array_type);
+      return nullptr;
+  }
+
+  // Double check that width matches TF data type size:
+  if (width != TF_DataTypeSize(dtype)) {
+    NAPI_THROW_ERROR(env, "Byte size of elements differs between JS VM and TF");
+    return nullptr;
+  }
+
+  // Determine the size of the buffer based on the dimensions.
+  size_t num_elements = 1;
+  for (size_t i = 0; i < shape_length; i++) {
+    num_elements *= shape[i];
+  }
+
+  // Ensure the shape matches the length of the passed in typed-array.
+  if (num_elements != array_length) {
+    NAPI_THROW_ERROR(env, "Shape does not match typed-array in bindData()");
+    return nullptr;
+  }
+
+  // Allocate and memcpy JS data to Tensor.
+  const size_t byte_size = num_elements * width;
+  TF_AutoTensor tensor(
+      TF_AllocateTensor(dtype, shape, shape_length, byte_size));
+  memcpy(TF_TensorData(tensor.tensor), array_data, byte_size);
```

Comment:
Travis complaining `memcpy` was not declared in this scope

Reply:
Ah thanks - it's defined in <cstring> updated.